src/ruby/ext/grpc/rb_channel_credentials.c

```diff
@@ -124,6 +125,84 @@ static ID id_pem_private_key;
 /* The attribute used on the mark object to hold the pem_private_key. */
 static ID id_pem_cert_chain;
 
+/* The attribute used on the mark object to hold the checkServerIdentity
+ * callback. */
+static ID id_check_server_identity_cb;
+
+struct verify_callback_params {
+  VALUE cb;
+  const char* servername;
+  const char* cert;
+};
+
+static VALUE verify_peer_callback_try_wrapper(VALUE arg) {
+  VALUE cb;
+  VALUE servername;
+  VALUE cert;
+
+  cb = rb_ary_entry(arg, 0);
+  servername = rb_ary_entry(arg, 1);
+  cert = rb_ary_entry(arg, 2);
+
+  if (rb_class_of(cb) == rb_cProc) {
+    rb_funcall(cb, rb_intern("call"), 2, servername, cert);
+  } else if (rb_class_of(cb) == rb_cSymbol) {
+    rb_funcall(rb_class_of(cb), rb_to_id(cb), 2, servername, cert);
+  } else {
+    printf(
+        "Callback argument in verify_peer_callback_try_wrapper is an invalid "
+        "type!\n");
+    return INT2NUM(1);
+  }
+  return INT2NUM(0);
+}
+
+static VALUE verify_peer_callback_catch_wrapper(VALUE arg,
+                                                VALUE exception_object) {
+  // Catch just always returns a failure signal.
+  return INT2NUM(1);
+}
+
+/* Before we jump back from native code (which doesn't have the GVL), it's
+   important to re-acquire it otherwise badness can happen. So this method
+   should be invoked with the GVL (i.e. by using the rb_thread_call_with_gvl()
+   method). */
+static void* invoke_rb_verify_callback_with_gvl(void* arg) {
+  VALUE result;
+  VALUE passthrough;
+  struct verify_callback_params* params = (struct verify_callback_params*)arg;
+
+  passthrough = rb_ary_new();
+  rb_ary_store(passthrough, 0, params->cb);
+  rb_ary_store(
+      passthrough, 1,
+      params->servername != NULL ? rb_str_new2(params->servername) : Qnil);
+  rb_ary_store(passthrough, 2,
+               params->cert != NULL ? rb_str_new2(params->cert) : Qnil);
+
+  result = rb_rescue(verify_peer_callback_try_wrapper, passthrough,
+                     verify_peer_callback_catch_wrapper, Qnil);
+  return NUM2INT(result) == 0 ? NULL : arg;
+}
+
+static int verify_peer_callback_wrapper(const char* servername,
+                                        const char* cert, void* userdata) {
+  struct verify_callback_params params;
+  if (userdata == NULL) {
+    printf("Error! Callback function wasn't set!\n");
+    return 1;
+  }
+
+  params.cb = (VALUE)userdata;
+  params.servername = servername;
+  params.cert = cert;
+
+  return rb_thread_call_with_gvl(invoke_rb_verify_callback_with_gvl, &params) ==
```

Comment:
I think the way this callback is invoked with respect to concurrency needs to be changed.

The C-Core is completely non-blocking and asynchronous. Since some C-Core thread (which may or may not be a ruby thread), is directly acquiring the interpreter lock and calling in to this ruby application handler, this has the potential to indefinitely block a C-Core thread and prevent intended work from being done.

One fix for this could be to dedicate a ruby thread to the task of invoking these callbacks. Such a thread could poll for work on a queue of "ready-to-run" "veroify peer callbacks". The actual verify peer callback handler that the C-Core invokes would just add to that queue, signal conditional variable, etc.. We already have such a thread for the task is invoking ruby-defined "call credentials" callbacks (See https://github.com/grpc/grpc/blob/master/src/ruby/ext/grpc/rb_event_thread.c#L126), so I think that adding another would could be reasonable. An optimization could be to combine both of these thread in to one, but I'm not sure that's worth doing.

Reply:
^ on second though, it might actually be simpler to combiner invocation of these handlers in to the call credentials callback thread, by using e.g. a "queue node tag type", but I'm not sure

Reply:
^^^ Oooh... good point!

Reply:
Also, just to add, granted an indefinitely blocking application callback would also deadlock a dedicated callback thread and prevent more callbacks from being fired, but IMO overall the dedicated thread is simpler; it's difficult to reason about what happens when a c-core thread acquires a ruby lock and runs ruby code.

Reply:
Hmm, what you're saying makes total sense, but the verify callback exposed by C-Core at present expects the callback to be synchronous so it doesn't pass in the on_peer_checked closure that would allow the callback to be asynchronous.
We could tweak the C-Core API so that callbacks are asynchronous? Which sounds like it might be important given interpreter locks. Is there anyone that should way in before I go down the route of changing that core API?

Reply:
cc @jiangtaoli2016 @markdroth for this core API change proposal (to allow the verify peer callback to be async)

Reply:
I agree, it's not okay to block in a callback from a C-core thread.

I'm not thrilled about adding yet another case where we need an asynchronous callback from C-core to the application.  We already have a few of these, and they've caused a number of headaches over time; it turns out to be fairly tricky to get the synchronization and state machinery right when leaving and re-entering C-core.

That having been said, I don't have a better suggestion, and the C-core API is not a public API, so we do have the flexibility to change this later if we can come up with something better.  So I'd say go ahead and add the callback to the API.

I suggest that you model the changes after one of the existing implementations.  The best one to look at is probably the metadata credentials plugin, whose API is defined here:

https://github.com/grpc/grpc/blob/master/include/grpc/grpc_security.h#L304

The implementation is here:

https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/plugin/plugin_credentials.cc
https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/plugin/plugin_credentials.h

Note that this API allows both sync and async returns.  The sync return is much simpler and more efficient in the case where no blocking is needed; the async return requires a callback and a thread hop, so it's more complex and less efficient.

One thing that is not handled correctly in the metadata credentials plugin API is cancellation.  Unfortunately, this API didn't originally support cancellation, and all of the wrapped language APIs were added without cancellation support.  When we later realized that we needed to support this, we were able to change the C-core API, but it was too late to fix the wrapped language APIs.  So we did a poor man's approach: we have C-core mark the request as cancelled, and when the wrapped language eventually invokes the callback, nothing actually happens.  But this means that we still wind up doing a lot of processing in the application that isn't actually necessary, because we have no way to inform it that the work is no longer needed.  I highly recommend not making that mistake with this new API: let's make sure that we expose a way of propagating cancelations to the wrapped language, so that this can be handled as a first-class citizen.

Please let me know if you have any questions.

Reply:
Thanks for all the pointers! I'll work on putting together an update to the core API and put that in a new PR. I'll come back to the python/ruby bindings in this PR once we get that ironed out. It'll probably be a little bit before I can get through that, so I'll flag you on it when it's ready.